godotproject/Player.gd

```diff
@@ -0,0 +1,185 @@
+extends KinematicBody2D
+
+signal tongue_start
+signal tongue_stop
+
+export (int) var speed = 200
+export (int) var jump_speed = -500
+export (int) var gravity = 1200
+export (float) var jump_length = 10 / 60.0
+export (float) var shorthop_length = 6 / 60.0
+export (float) var shorthop_ratio = 0.6
+export (float) var swing_user_anglespeed_per_s = 1
+export (float) var swing_user_radius_per_s = 40
+
+
+var velocity = Vector2.ZERO
+var jump_frame_countdown = 0
+var jump_shorthop_countdown = 0
+var facing = Vector2.RIGHT
+
+var user_direction = Vector2.ZERO
+var jump_pressed = false
+var tongue_pressed = false
+var jump_held = false
+var tongue_held = false
+
+var swing_angular_speed = 0
+var swing_radius = 0
+var swing_angle = 0
+var swing_pivot_position = Vector2.ZERO
+
+func start_swing():
+	# Calculate initial angle and radius
+	var swing_vec = update_swing_radius_angle()
+	var swing_dir_vec = swing_vec.rotated(PI/2)
+	var dotprod = swing_dir_vec.dot(velocity)
+	if abs(dotprod)<  0.01 or swing_radius < 0.01:
+		swing_angular_speed = 0
+	else:
+		swing_angular_speed = dotprod / abs(dotprod) * pow(abs(dotprod), 0.5) / max(1, swing_radius)
+
+func stop_swing():
+	# Convert swinging angular momentum into player velocity
+	velocity = Vector2.DOWN.rotated(swing_angle) * swing_angular_speed * swing_radius
+	print_debug('Stop swing: Velocity=',velocity)
+	emit_signal("tongue_stop")
+
+func update_swing_radius_angle():
+	# Calculate current angle and radius
+	var frog_pos = get_global_transform().xform(Vector2.ZERO)
+	var swing_vec = frog_pos - swing_pivot_position
+	swing_radius = swing_vec.length()
+	swing_angle = swing_vec.angle() # Angle from positive X-axis to swing vector
+	return swing_vec
+
+
+func get_input():
+	var l = Input.is_action_pressed("walk_left")
+	var r = Input.is_action_pressed("walk_right")
+	var u = Input.is_action_pressed("look_up")
+	var d = Input.is_action_pressed("look_down")
+	user_direction.x = 0
+	if l:
+		user_direction.x -= 1
+	if r:
+		user_direction.x += 1
+	user_direction.y = 0
+	if u:
+		user_direction.y -= 1
+	if d:
+		user_direction.y += 1
+	jump_pressed = Input.is_action_just_pressed("jump")
+	jump_held = Input.is_action_pressed("jump")
+	tongue_pressed = Input.is_action_just_pressed("tongue")
+	tongue_held = Input.is_action_pressed("tongue")
+
+func _draw():
+	# THIS IS JUST DEBUG
+	# TODO: Remove
+	# Draw swinging velocity vector
+	if $PlayerTongue.swinging:
+		var tgt = Vector2.DOWN.rotated(swing_angle) * swing_angular_speed * 100
+		draw_line(Vector2(0,0), tgt, Color(1,0,0))
+
+func do_movement(delta):
+	if $PlayerTongue.swinging:
+		# Handle user input
+		swing_angular_speed += swing_user_anglespeed_per_s * -user_direction.x * delta
+		# TODO: Fix this and uncomment
```

Comment:
fix and uncomment? If its going to take a bit of time then we can leave it in and update in a future PR